### ISSUE TITLE
Type inference fails with nested arrow function callback with block

### ISSUE THREAD

**Issue by Rugvip** [microsoft/TypeScript#50687]
# Bug Report

### 🔎 Search Terms

callback, block, inference, double

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about arrow functions

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAJwKYAdlwDwBUB8AFAPoQLAwDmAXIgN4BQii6AhsqwLYDOtuA3E0QRWAG1EAjVhADWtQmw48+ASkQBefIgXsuvRLjWbEYEJwmpkggL4raANzgwAJvRSooIZEmsMGaTDhCRmZFPVoARgAaIRFxKVl5Iy06d09vRFYNLSyAakQIxGsoxAB6UsRwGTA4AHcwBltBAKxgoTDlApjmOMlpOURiCWy3NHSkLOM8gqKS8sQEUQBPRGcYYGBLVEhURBhuRCgAC12qGrRXDsaVZoxWkJZdTujYsT7E7WTMkenosorFis1hstjs9gdjrtRP0FsBEAADVJjLxIAB06KK8Ou-CAA

### 💻 Code

```ts
function repro<T>(_config: {
  params: T;
  callback: (params: T) => (params: T) => number;
}): void { return }

repro({
  params: 1,
  callback: () => { return a => a + 1 }, // a is unknown
});
repro({
  params: 1,
  callback: _b => { return a => a + 1 }, // a is a number, only difference is the ignored outer param
});
repro({
  params: 1,
  callback: () => a => a + 1, // a is a number, only difference is the lack of `{ return ... }`
});
```

### 🙁 Actual behavior

The type of `a` ends up being `unknown`, but only if no parameter is declared and a block is used in the outer callback.

### 🙂 Expected behavior

The type of the inner callback parameter `a` should be known in all 3 cases.

**Comment by Urtgard** [comment]
In my project I stumbled over a similiar case:
```ts
class Test {
  array = [1, 2, 3]
  optional?: number

  public methodError(): void {
    if (!this.optional) return;

    this.array.forEach((x) => {
      x = this.optional; // Type 'number | undefined' is not assignable to type 'number'. Type 'undefined' is not assignable to type 'number'.(2322)
    });
  }

  public methodOk(): void {
    this.array.forEach((x) => {
      if (!this.optional) return;
      x = this.optional;
    });
  }
}
```
Let me know if this is different problem, then I'll open a new issue.

**Comment by Andarist** [comment]
@Urtgard your problem is different and it's a not a bug, take a look at the inline explanation that I have added in this [TS playground](https://www.typescriptlang.org/play?#code/MYGwhgzhAEAqCmEAu0DeAoa0wCcdgE9oBeaAbQEYAaaAJhoGYBdAbk2gHsAHJASw4B2YEAH4AXNAEBXALYAjeDjbsuUuSF7BoM+EgAWHACYBRPBxwAKAJQSAbh16G07LLwBm0CwEJ9vCADpuPkFhK2gcXSkcAWUsLAB6eM4AaxoFYDApCHhoAHcc3BzkXhAQaF4BbGgIYG4c3L1FHIADXwCg-iEQZug9SGgFeEqhM3ynQw5cgRdoNv9cfAIyAAYmElm9P0CeTuFYuLmFwn83c2MwYD0LCwAPMOIAPmc4uMTJDhQOVIH4DKychRIfJDDY5DhSJCKaq1LgFARONrQDKlOQXZLQVqbdo7ELdJHgkBOPq2AHwEERSAQXgAcwE8EMMwSSQq0C4+GAfGALVOOHOlx6yJAMH0OUFqOA6IgBAElxwgnBEBARDkELgAGVoBNEAIAOQoKTwxTIMDwjZgJCMjZbI5LVbrOYdXFsF4AXyszugLvQXqAA)

**Comment by nix6839** [comment]
```ts
function fn<ToInferred>(
  callback: (unused: string) => {
    key: ToInferred;
    keyCallback: (key: ToInferred) => number;
  },
) {}

fn(() => ({
  key: 5,
  keyCallback: (key) => {
    return key;
  },
}));

fn((unused) => ({
  key: 5,
  keyCallback: (key) => {
    return key; // key is unknown.
  },
}));
```
Same issue here... I hope it will be fixed at v4.9.2
